buildSrc/src/main/groovy/io/micronaut/build/internal/WriteMicronautVersionInfoTask.java

```diff
@@ -0,0 +1,102 @@
+package io.micronaut.build.internal;
+
+import groovy.lang.Closure;
+import groovy.xml.XmlSlurper;
+import groovy.xml.slurpersupport.GPathResult;
+import groovy.xml.slurpersupport.NodeChild;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.ResolvedArtifact;
+import org.gradle.api.artifacts.ResolvedDependency;
+import org.gradle.api.artifacts.result.ArtifactResolutionResult;
+import org.gradle.api.artifacts.result.ComponentArtifactsResult;
+import org.gradle.api.artifacts.result.ResolvedArtifactResult;
+import org.gradle.api.file.DirectoryProperty;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.CacheableTask;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.maven.MavenModule;
+import org.gradle.maven.MavenPomArtifact;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.ParserConfigurationException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.TreeMap;
+
+@CacheableTask
+public abstract class WriteMicronautVersionInfoTask extends DefaultTask {
+
+    public static final String MICRONAUT_VERSIONS_PROPERTIES_FILE_NAME = "micronaut-versions.properties";
+
+    @Input
+    public abstract Property<String> getVersion();
+
+    @Input
+    public abstract ListProperty<String> getExtraBomProperties();
+
+    @OutputDirectory
+    public abstract DirectoryProperty getOutputDirectory();
+
+    public WriteMicronautVersionInfoTask() {
+        getOutputs().doNotCacheIf("snapshot version", spec -> getVersion().get().endsWith("SNAPSHOT"));
```

Comment:
Why do we need this? This hints at a missing task input.

Reply:
I brought it across from the task in starter... I will have a think

Reply:
@melix This seems to work, and feels much nicer

https://github.com/micronaut-projects/micronaut-core/pull/7202/commits/e595a35095cc72f673cdcdf6a44ce613a293e1a5

PTAL, I may be committing Gradle heresy 😉

Reply:
And then cleaned up the old input in a second commit 🙄. I got too excited

Reply:
I think you should use a `Configuration` directly, not a `Property<Configuration>`, because that's a `FileCollection` already. As I understood this isn't configuration cache compatible, but we're far from there I think (cc @eskatos)

Reply:
This is because you need `@InputFiles`

Reply:
Gotcha 👍

Hopefully this does it 🤞 https://github.com/micronaut-projects/micronaut-core/pull/7202/commits/0cc5b8f58b48dadb0b23b624a7a3e0eae17ded84

Safe travels!

Reply:
I'm late to the party but here are some thoughts.

Yep, `Configuration` is a lazy type already so no need to wrap it in a `Property`.

You're right Cedric, using a `Configuration` input is not compatible with configuration caching. `Configuration` objects are way too heavyweight.

IIUC the inputs of the task are the _declared_ dependencies, correct?

Reply:
Hiya 👋😄

Yeah, I need to parse the BOM dependencies' POM to extract out the properties

So i was stick the boms as dependencies in a custom configuration

https://github.com/micronaut-projects/micronaut-core/pull/7202/files#diff-03d6884a03b4390e72895a091b6c223d2d143290ddc4118f5a9a5354753524aeR15

(There's only 1 at the moment, but it may be more later)

Reply:
The inputs are actually neither the artifacts (jars) nor the _declared dependencies_, but metadata files. We can _simplify_ to declared dependencies probably, as long as the repositories we resolve from are assumed constant.

Reply:
Do you have a pointer to what you mean?

Reply:
Do you mean as we do in starter?  https://github.com/micronaut-projects/micronaut-starter/pull/1176/files

We should probably align how we do this so it looks like one person wrote both 😔

Reply:
> Do you have a pointer to what you mean?

Sorry for not being clear enough. What I meant is that _technically speaking_, the inputs of this task are the metadata files (pom files, and possibly their parent POMs). This is what should be declared as an input. In practice there's no easy way to declare this. So we _could_ simplify by saying that the inputs of the task are the declared dependencies, and let the task resolve the pom files. Doing so assumes that we would always download the pom files from the same location (so that we can guarantee that they are always the same pom files for a given dependency declaration). If they are not, then the repositories are _also_ an input.

There's no specific example for this, what we do now is probably enough: we're going to resolve the files, which is a bit too much but no big deal.